packages/propeller/src/StreamsSink.mts

```diff
@@ -16,38 +16,117 @@ function groupByStream(items: [StreamName, ITimelineEvent][]): Map<StreamName, I
+      this.events.push(event)
+      return
+    }
+    if (last.index > event.index) return
+    if (event.index > last.index) throw new Error("Unexpected gap")
+  }
+}
+
+const getOrAdd = <K, V>(map: Map<K, V>, key: K, val: () => V) => {
+  const current = map.get(key)
+  if (current == null) {
+    const value = val()
+    map.set(key, value)
+    return value
+  }
+  return current
+}
+
 export class StreamsSink implements Sink {
-  private readonly limit: LimitFunction
+  private queue = new Queue<Stream>()
+  private streams = new Map<StreamName, Stream>()
+  private batchStreams = new Map<IngesterBatch, Set<StreamName>>()
+  private onReady?: () => void 
+  private inProgressBatches = 0
   constructor(
     private readonly handle: EventHandler<string>,
-    maxConcurrentStreams: number,
+    private maxConcurrentStreams: number,
+    private maxReadAhead: number,
     private readonly tracingAttrs: Attributes = {},
   ) {
-    this.limit = pLimit(maxConcurrentStreams)
     this.addTracingAttrs({ "eqx.max_concurrent_streams": maxConcurrentStreams })
   }
 
   addTracingAttrs(attrs: Attributes) {
     Object.assign(this.tracingAttrs, attrs)
   }
 
-  async pump(batch: Batch) {
-    const streams = groupByStream(batch.items)
-    const promises: Promise<void>[] = []
-    for (const [streamName, events] of streams) {
-      promises.push(
-        this.limit(() =>
-          traceHandler(
-            tracer,
-            this.tracingAttrs,
-            StreamName.category(streamName),
-            streamName,
-            events,
-            this.handle,
-          ),
-        ),
-      )
-    }
-    await Promise.all(promises)
+  start(signal: AbortSignal) {
+    return new Promise<void>((_resolve, reject) => {
+      let stopped = false
+      const aux = async (): Promise<void> => {
+        if (signal.aborted || stopped) return
+        const stream = this.queue.tryGet()
+        if (!stream) return void setImmediate(aux)
+        this.streams.delete(stream.name)
+        await traceHandler(
+          tracer,
+          this.tracingAttrs,
+          StreamName.category(stream.name),
+          stream.name,
+          stream.events,
+          this.handle,
+        ).catch((err) => {
+          reject(err)
+          stopped = true
+        })
+        for (const [batch, streams] of this.batchStreams) {
+          streams.delete(stream.name)
+          if (streams.size === 0) {
+            batch.onComplete()
+            this.batchStreams.delete(batch)
+            this.inProgressBatches--
+            this.onReady?.()
+            delete this.onReady
+          }
+        }
+        setImmediate(aux)
+      }
+
+      for (let i = 0; i < this.maxConcurrentStreams; ++i) aux()
+    })
+  }
+
+  pump(batch: IngesterBatch, signal: AbortSignal) {
+    return new Promise<void>((resolve, reject) => {
+      this.inProgressBatches++
+      const names = new Set<StreamName>()
+      this.batchStreams.set(batch, names)
```

Comment:
The thought process here is:

The queue is processed in order, and streams are added to the queue in batch-order.
We mutate the value of the stream in the queue when more batches come in with data for the same stream.
When we start processing a stream we remove it from the map, meaning that if another batch comes in for the same stream it'll add another `Stream` to the queue and map.

I think you're right that there's an edge case, but I think it's in the opposite direction. That is, it's possible to commit too early because we can remove the stream from the set before having actually handled all the events from that batch.

I think the solution is to use a set of `Stream` instead of `StreamName` 🤔

Reply:
Let's get on a call soon so I can take you through the logic 😅

Reply:
it boils down to on line 88 we go through each batch and handle the fact that the stream is processed. This means removing it from the set, and if the set becomes empty as a result completing the batch

Reply:
I'll have to check the spec to see if the iteration order of a map is stable and in insertion order, if not I'll move to a ResizeArray

Reply:
Seems correct in general
Things are def simpler if handlers don't get to say "used half of them" and/or I failed; chuck them back in the pool for retry (merged with any that arived in the interim)
call won't fit today unfortunately

Reply:
Was definitely not intending “soon” to mean today 😅